primitives/src/merkle_tree/light_weight.rs

```diff
@@ -0,0 +1,195 @@
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! A light weight merkle tree is an append only merkle tree who only keeps its
+//! frontier -- the right-most path.
+
+use core::ops::AddAssign;
+
+use super::{
+    internal::{build_light_weight_tree_internal, MerkleNode, MerkleProof, MerkleTreeCommitment},
+    AppendableMerkleTreeScheme, DigestAlgorithm, Element, ForgetableMerkleTreeScheme, Index,
+    LookupResult, MerkleCommitment, MerkleTreeScheme, NodeValue, ToTraversalPath,
+};
+use crate::{errors::PrimitivesError, impl_forgetable_merkle_tree_scheme, impl_merkle_tree_scheme};
+use ark_std::{
+    borrow::Borrow, boxed::Box, fmt::Debug, marker::PhantomData, string::ToString, vec, vec::Vec,
+};
+use num_bigint::BigUint;
+use num_traits::pow::pow;
+use serde::{Deserialize, Serialize};
+use typenum::Unsigned;
+
+impl_merkle_tree_scheme!(LightWeightMerkleTree, build_light_weight_tree_internal);
+impl_forgetable_merkle_tree_scheme!(LightWeightMerkleTree);
+
+impl<E, H, I, Arity, T> AppendableMerkleTreeScheme for LightWeightMerkleTree<E, H, I, Arity, T>
+where
+    E: Element,
+    H: DigestAlgorithm<E, I, T>,
+    I: Index + From<u64> + AddAssign + ToTraversalPath<Arity>,
+    Arity: Unsigned,
+    T: NodeValue,
+{
+    fn push(&mut self, elem: impl Borrow<Self::Element>) -> Result<(), PrimitivesError> {
+        <Self as AppendableMerkleTreeScheme>::extend(self, [elem])
+    }
+
+    fn extend(
+        &mut self,
+        elems: impl IntoIterator<Item = impl Borrow<Self::Element>>,
+    ) -> Result<(), PrimitivesError> {
+        let mut iter = elems.into_iter().peekable();
+
+        let traversal_path = I::from(self.num_leaves).to_traversal_path(self.height);
+        self.num_leaves += self.root.extend_and_forget_internal::<H, Arity>(
+            self.height,
+            &I::from(self.num_leaves),
+            &traversal_path,
+            true,
+            &mut iter,
+        )?;
+        if iter.peek().is_some() {
+            return Err(PrimitivesError::ParameterError(
+                "Exceed merkle tree capacity".to_string(),
+            ));
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod mt_tests {
+    use crate::{
+        merkle_tree::{
+            internal::{MerkleNode, MerkleProof},
+            prelude::{RescueLightWeightMerkleTree, RescueMerkleTree},
+            *,
+        },
+        rescue::RescueParameter,
+    };
+    use ark_ed_on_bls12_377::Fq as Fq377;
+    use ark_ed_on_bls12_381::Fq as Fq381;
+    use ark_ed_on_bn254::Fq as Fq254;
+
+    #[test]
+    fn test_light_mt_builder() {
+        test_light_mt_builder_helper::<Fq254>();
+        test_light_mt_builder_helper::<Fq377>();
+        test_light_mt_builder_helper::<Fq381>();
+    }
+
+    fn test_light_mt_builder_helper<F: RescueParameter>() {
+        assert!(RescueLightWeightMerkleTree::<F>::from_elems(1, &[F::from(0u64); 3]).is_ok());
```

Comment:
Maybe make `3` a parameter (e.g. `let arity = RescueLightWeightMerkleTree::ARITY`)? Then replace the numbers below to be expressions related to `arity`.

The advantage is that when we change the setup of `RescueLightWeightMerkleTree`, the tests won't be broken.

Reply:
Good point, will do.